src/box/relay.cc

```diff
@@ -785,8 +788,16 @@ relay_send_heartbeat(struct relay *relay)
 		++relay->last_sent_ack.vclock_sync;
 		RegionGuard region_guard(&fiber()->gc);
 		xrow_encode_relay_heartbeat(&row, &relay->last_sent_ack);
-		row.tm = ev_now(loop());
+		/*
+		 * Do not encode timestamp if this heartbeat is sent in between
+		 * data rows so as to not affect replica's upstream lag.
+		 */
+		if (relay->last_row_time > relay->last_heartbeat_time)
+			row.tm = 0;
+		else
+			row.tm = ev_now(loop());
```

Comment:
Maybe not encode it for heartbeats ever? Why is it even needed for non-data rows?

Reply:
It's needed to notify the replica that it's up to date (synced).
If the master has no data to send, it sends a heartbeat with a non-zero row.tm, and upon receipt replica updates 'upstream.lag' and possibly transitions from 'sync' to 'follow'.

Besides, when there is no data,  upstream lag measures the network lag between the two nodes (if we ignore the fact that their clocks might diverge).

I'd rather keep it as is for now. Even though having 2 usages for 'upstream lag' is strange (the first usage is "fow far behind we are, measured in seconds" and the other is "network lag")